### ISSUE TITLE
Update cmake minimum version

### ISSUE THREAD

**Issue by bartgol** [E3SM-Project/EKAT#324]
The current minimum version is 3.3, but recent cmake (3.27) started to give this warning
```
CMake Deprecation Warning at /home/lbertag/workdir/scream/scream-src/master/externals/ekat/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.
```
Considering that most machines feature a relatively recent  cmake, we could probably bump this up to something like 3.1X. [Here](https://cmake.org/cmake/help/latest/release/index.html) you can find the release notes for 3.X. Some interesting ones that may be relevant for ekat:

- 3.11: adds [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html#fetchcontent): a somewhat hybrid approach between `add_subdirectory` and `ExternalProject_add`.
- 3.12: `find_package` searches `<PackageName>_ROOT` cmake/env variables.
- 3.14: [FetchContent_MakeAvailable](https://cmake.org/cmake/help/latest/module/FetchContent.html#command:fetchcontent_makeavailable) is added
- 3.15: [cmake_parse_arguments()](https://cmake.org/cmake/help/latest/command/cmake_parse_arguments.html#command:cmake_parse_arguments) adds a way to check if a keyword arg is missing values.
- 3.16: [target_precompile_headers()](https://cmake.org/cmake/help/latest/command/target_precompile_headers.html#command:target_precompile_headers)

In general, each version also adds QoL improvements, (see the `ZIP_LISTS` option in `foreach` added in 3.17), so the more recent the version, the more robust/concise our CMake syntax can be. Of course, asking for 3.28 is too much (I don't have it on my laptop either), so we need to be a bit conservative. Nevertheless, 3.3 is a bit old...

@jgfouca @jeff-cohere @tcclevenger thoughts?

**Comment by jeff-cohere** [comment]
I think we should use at least 3.12 to avoid violating the Principle of Least Astonishment for using `find_package`.

**Comment by bartgol** [comment]
Yes. The `<PackageName>_ROOT` is something quite useful, since it allows to avoid having to add all those `HINTS`/`PATHS` args to the `find_package` calls. In a way, it's a _standardized_ way of specifying `PATHS` and `HINTS`.

**Comment by jgfouca** [comment]
All the TPL work I did a few months ago depends on `<PackageName>_ROOT` being available, so at the very least 3.12 should be the minimum.

**Comment by bartgol** [comment]
And it's fair to assume at this point that 3.12 is available on all machines. CMake v3.12.0 was released in Nov 2018...
